langchain/llms/gpt4all.py

```diff
@@ -0,0 +1,182 @@
+
+    use_mlock: bool = Field(False, alias="use_mlock")
+    """Force system to keep model in RAM."""
+
+    embedding: bool = Field(False, alias="embedding")
+    """Use embedding mode only."""
+
+    n_threads: Optional[int] = Field(4, alias="n_threads")
+    """Number of threads to use."""
+
+    n_predict: Optional[int] = 256
+    """The maximum number of tokens to generate."""
+
+    temp: Optional[float] = 0.8
+    """The temperature to use for sampling."""
+
+    top_p: Optional[float] = 0.95
+    """The top-p value to use for sampling."""
+
+    top_k: Optional[int] = 40
+    """The top-k value to use for sampling."""
+
+    echo: Optional[bool] = False
+    """Whether to echo the prompt."""
+
+    stop: Optional[List[str]] = []
+    """A list of strings to stop generation when encountered."""
+
+    repeat_last_n: Optional[int] = 64
+    "Last n tokens to penalize"
+
+    repeat_penalty: Optional[float] = 1.3
+    """The penalty to apply to repeated tokens."""
+
+    n_batch: int = Field(1, alias="n_batch")
+    """Batch size for prompt processing."""
+
+    streaming: bool = False
+    """Whether to stream the results or not."""
+
+    client: Any = None  #: :meta private:
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+
+    @property
+    def _default_params(self) -> Dict[str, Any]:
+        """Get the identifying parameters."""
+        return {
+            "seed": self.seed,
+            "n_predict": self.n_predict,
+            "n_threads": self.n_threads,
+            "n_batch": self.n_batch,
+            "repeat_last_n": self.repeat_last_n,
+            "repeat_penalty": self.repeat_penalty,
+            "top_k": self.top_k,
+            "top_p": self.top_p,
+            "temp": self.temp,
+        }
+
+    @staticmethod
+    def _llama_param_names() -> Set[str]:
+        """Get the identifying parameters."""
+        return {
+            "seed",
+            "n_ctx",
+            "n_parts",
+            "f16_kv",
+            "logits_all",
+            "vocab_only",
+            "use_mlock",
+            "embedding",
+        }
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        """Validate that the python package exists in the environment."""
+        try:
+            from pyllamacpp.model import Model as GPT4AllModel
+
+            llama_keys = cls._llama_param_names()
+            model_kwargs = {k: v for k, v in values.items() if k in llama_keys}
+            values["client"] = GPT4AllModel(
+                ggml_model=values["model"],
+                **model_kwargs,
+            )
+
+        except ImportError:
+            raise ValueError(
+                "Could not import pyllamacpp python package. "
+                "Please install it with `pip install pyllamacpp`."
+            )
+        return values
+
+    @property
+    def _identifying_params(self) -> Mapping[str, Any]:
```

Comment:
Is usage of immutable on return type (and mutables on input like for `stop` token) intentional?

(I tend to flip it around)

Reply:
not really tbh. i dont think this is done with the intention of being immutable on return type, but rather to expose a helper internal method